File: 18-4sum/4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        ans=[]
        nums.sort()
        for j in range(len(nums)-3):
            if j>0 and nums[j]==nums[j-1]:
                continue
            for i in range(j+1,len(nums)-2):
                if i>j+1 and nums[i]==nums[i-1]:
                    continue
                lo=i+1
                hi=len(nums)-1
                while lo<hi:
                    if nums[j]+nums[i]+nums[lo]+nums[hi]==target:
                        ans.append([nums[j],nums[i],nums[lo],nums[hi]])
                        lo=lo+1
                        hi=hi-1
                        while lo<hi and nums[lo]==nums[lo-1]:lo=lo+1
                        while lo<hi and nums[hi]==nums[hi+1]:hi=hi-1
                    elif nums[j]+nums[i]+nums[lo]+nums[hi]>target:
                        hi=hi-1
                    else:
                        lo=lo+1
        return ans
        

File: 18-4sum/4sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        int length=nums.length;

        List<List<Integer>> ans=new ArrayList<>();

        for(int i=0;i<length-3;i++){

            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }

            for(int j=i+1;j<length-2;j++){
                if(j>i+1 && nums[j]==nums[j-1]){
                    continue;
                }

                int low=j+1;
                int high=length-1;

                while(low<high){
                    long sum=(long) nums[i]+nums[j]+nums[low]+nums[high];
                    if(sum==target){
                        List<Integer> list=new ArrayList<>();
                        list.add(nums[i]);
                        list.add(nums[j]);
                        list.add(nums[low]);
                        list.add(nums[high]);
                        ans.add(list);
                        low=low+1;
                        high=high-1;
                        while(low<high && nums[low]==nums[low-1]){
                            low=low+1;
                        }
                        while(low<high && nums[high]==nums[high+1]){
                            high=high-1;
                        }
                    }else if(sum<target){
                        low=low+1;
                    }else{
                        high=high-1;
                    }
                }
            }
        }

        return ans;
        
    }
}